### PR TITLE
Fixed #4784 : Inform users about unsupported features in beta flavor

### LINKED ISSUE #4784: "This feature is only available on the prod flavor."
Several features of the app are broken when you try to use them with the [beta flavor](https://github.com/commons-app/commons-app-documentation/blob/master/android/Quick-start-guide-for-Developers.md#building-the-source-code).
For instance, the [leaderboard throws an exception when running the beta flavor](https://github.com/commons-app/apps-android-commons/issues/4774#issuecomment-1027151359). It wastes everyone's time.

When opening the activity when such features are present, it would be helpful to show such a toast:

> This feature is only available on the prod flavor.

Activities for which the toast should be shown:
- Leaderboard
- Maybe "Review"?
- Maybe "Explore" > search for something > "Items"?
- If you find others, please comment :-)

_While our long-term goal should be to make all features work with the beta flavor too, for now we should at least clearly identify what features do not work._

### PR DESCRIPTION
**Pull request by KosOrKosm**
### **Description**

Fixes #4784 

### **What changes did you make and why?**

The app is receiving empty results or HTTP 500 responses when trying to retrieve the leaderboard or achievements for users in the Beta environment.

Added a toast that appears whenever AchievementsFragment or LeaderboardFragment is made visible on Beta flavors, informing the user these features are unavailable in Beta flavor. Also skips the dysfunctional API calls and hides the empty / dysfunctional UI elements.

### **Tests performed**

Tested betaDebug + prodDebug on Pixel 3a (AVD Emulator) with API level 28.

### **Screenshots**

image image

### CODE REVIEW

File: app/src/test/kotlin/fr/free/nrw/commons/leaderboard/LeaderboardFragmentUnitTests.kt [hunk 1]
```diff
@@ -200,4 +201,15 @@ class LeaderboardFragmentUnitTests {
         method.invoke(fragment)
     }
 
+    @Test
+    @Throws(Exception::class)
+    fun testMenuVisibilityOverrideNoContext() {
```
Comment (madhurgupta10) [inline app/src/test/kotlin/fr/free/nrw/commons/leaderboard/LeaderboardFragmentUnitTests.kt:206]: You can add assert equals here to check if Toast contains the right text

```
Assert.assertEquals(ShadowToast.getTextOfLatestToast().toString(), getString(R.string.leaderboard_unavailable_beta))
```

File: app/src/test/kotlin/fr/free/nrw/commons/profile/achievements/AchievementsFragmentUnitTests.kt [hunk 2]
```diff
@@ -326,4 +326,16 @@ class AchievementsFragmentUnitTests {
         method.isAccessible = true
         method.invoke(fragment)
     }
+
+    @Test
+    @Throws(Exception::class)
+    fun testMenuVisibilityOverrideNoContext() {
```
Comment (madhurgupta10) [inline app/src/test/kotlin/fr/free/nrw/commons/profile/achievements/AchievementsFragmentUnitTests.kt:332]: Similar to other test function add asset for toast string

File: app/src/main/java/fr/free/nrw/commons/profile/leaderboard/LeaderboardFragment.java [hunk 3]
```diff
@@ -174,6 +183,19 @@ public void onNothingSelected(AdapterView<?> adapterView) {
         return rootView;
     }
 
+    @Override
+    public void setMenuVisibility(boolean visible) {
+        super.setMenuVisibility(visible);
+
+        // Whenever this fragment is revealed in a menu,
+        // notify Beta users the page data is unavailable
+        if(ConfigUtils.isBetaFlavour() && visible && getContext() != null) {
```
Comment (madhurgupta10) [inline app/src/main/java/fr/free/nrw/commons/profile/leaderboard/LeaderboardFragment.java:192]: Add test when false is passed as the argument, in that case toast should not appear

File: app/src/main/java/fr/free/nrw/commons/profile/achievements/AchievementsFragment.java [hunk 4]
```diff
@@ -179,11 +181,38 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
...
+            setMenuVisibility(true);
+            return rootView;
+        }
         setWikidataEditCount();
         setAchievements();
         return rootView;
     }
 
+    @Override
+    public void setMenuVisibility(boolean visible) {
```
Comment (madhurgupta10) [inline app/src/main/java/fr/free/nrw/commons/profile/achievements/AchievementsFragment.java:204]: Add test when false is passed as the argument, in that case toast should not appear

**Review by madhurgupta10 — CHANGES_REQUESTED** [review]

**Comment by madhurgupta10** [comment]
@KosOrKosm Could you also sync your branch with the latest master?

**Comment by KosOrKosm** [comment]
@madhurgupta10 Synced with master and updated tests to check if the Toast appears.

**Review by madhurgupta10 — APPROVED** [review]
Haven't tested but the changes looks good to me.

**Comment by KosOrKosm** [comment]
Hello @nicolas-raoul, could you review/test my PR?

**Comment by nicolas-raoul** [comment]
Sorry for the delay!
Strangely at the moment I am struggling to get the beta flavor to run:
```
$ ./gradlew installBetaDebug
[...]
Installing APK 'app-commons-v3.1.1-KosOrKosm-beta-disable-prod-only-apis-beta-debug.apk' on 'SM-G970F - 9' for app:betaDebug
Installed on 1 device.
[...]
BUILD SUCCESSFUL in 3m 10s
$ adb shell am start -n fr.free.nrw.commons/fr.free.nrw.commonsCommonsApplication
Starting: Intent { cmp=fr.free.nrw.commons/fr.free.nrw.commonsCommonsApplication }
Error type 3
Error: Activity class {fr.free.nrw.commons/fr.free.nrw.commonsCommonsApplication} does not exist.
```
Any idea what might be wrong, by any chance? :-)

File: app/src/main/java/fr/free/nrw/commons/profile/achievements/AchievementsFragment.java [hunk 5]
```diff
@@ -179,11 +182,46 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
...
+            progressBar.setVisibility(View.GONE);
+            imageByWikiText.setText(R.string.no_image);
+            imageRevertedText.setText(R.string.no_image_reverted);
+            imageUploadedText.setText(R.string.no_image_uploaded);
+            wikidataEditsText.setText("0");
+            imagesFeatured.setText("0");
+            tvQualityImages.setText("0");
+            thanksReceived.setText("0");
+            setMenuVisibility(true);
+            return rootView;
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/profile/achievements/AchievementsFragment.java:197]: Would it be possible to add a unit test covering these lines?

**Comment by nicolas-raoul** [comment]
Pushing the APK and installing it directly worked, strange...

Anyway, I tested and it is working great! :-)

Would you mind just addressing my previous coverage comment?
Thanks!

**Comment by KosOrKosm** [comment]
@nicolas-raoul I added a test that should cover the missing lines in AchievementsFragment

**Review by nicolas-raoul — APPROVED** [review]

**Comment by nicolas-raoul** [comment]
This pull request unfortunately broke some unit tests:
```
fr.free.nrw.commons.leaderboard.LeaderboardFragmentUnitTests > testMenuVisibilityOverrideVisibleWithContext FAILED
    java.lang.Exception at AndroidTestEnvironment.java:538
        Caused by: java.lang.NullPointerException at LeaderboardFragmentUnitTests.kt:232

fr.free.nrw.commons.profile.achievements.AchievementsFragmentUnitTests > testMenuVisibilityOverrideVisibleWithContext FAILED
    java.lang.Exception at AndroidTestEnvironment.java:538
        Caused by: java.lang.NullPointerException at AchievementsFragmentUnitTests.kt:375
```

**Comment by KosOrKosm** [comment]
That's odd, the tests pass in my local environment. I think the test might be failing on production because of the beta flavor check, which causes the test to never display a Toast and the assertion to fail. Is there any way to overwrite the build flavor for the duration of a test?

**Comment by madhurgupta10** [comment]
> This pull request unfortunately broke some unit tests:
> 
> ```
> fr.free.nrw.commons.leaderboard.LeaderboardFragmentUnitTests > testMenuVisibilityOverrideVisibleWithContext FAILED
>     java.lang.Exception at AndroidTestEnvironment.java:538
>         Caused by: java.lang.NullPointerException at LeaderboardFragmentUnitTests.kt:232
> 
> fr.free.nrw.commons.profile.achievements.AchievementsFragmentUnitTests > testMenuVisibilityOverrideVisibleWithContext FAILED
>     java.lang.Exception at AndroidTestEnvironment.java:538
>         Caused by: java.lang.NullPointerException at AchievementsFragmentUnitTests.kt:375
> ```

The latest build (https://github.com/commons-app/apps-android-commons/runs/5512038434?check_suite_focus=true) seems to be successful without any failing tests, @nicolas-raoul can you please confirm when you detected these failing tests?

**Comment by nicolas-raoul** [comment]
@madhurgupta10 I had found this commit using `git bissect` running `./gradlew testProdDebugUnitTest`, but I might have messed up something. Anyway master tests are passing now, thanks all! :-)

**Pull request merged**